flexmeasures/utils/calculations.py

```diff
@@ -37,11 +38,51 @@ def drop_nan_rows(a, b):
     return d[:, 0], d[:, 1]
 
 
+def apply_stock_changes_and_losses(
+    initial: float,
+    changes: list[float],
+    storage_efficiency: float | list[float],
+    how: str = "linear",
+    decimal_precision: int | None = None,
+) -> list[float]:
+    """Assign stock changes and determine losses from storage efficiency.
+
+    :param initial:             initial stock
+    :param changes:             stock change for each step
+    :param storage_efficiency:  ratio of stock left after a step (constant ratio or one per step)
+    :param how:                 left, right or linear; how stock changes should be applied, which affects how losses are applied
+    :param decimal_precision:   Optional decimal precision to round off results (useful for tests failing over machine precision)
+    """
+    stocks = [initial]
+    if not isinstance(storage_efficiency, list):
+        storage_efficiency = [storage_efficiency] * len(changes)
+    for d, e in zip(changes, storage_efficiency):
+        s = stocks[-1]
+        if e == 1:
+            next_stock = s + d
+        elif how == "left":
+            # First apply the stock change, then apply the losses (i.e. the stock changes on the left side of the time interval in which the losses apply)
+            next_stock = (s + d) * e
+        elif how == "right":
+            # First apply the losses, then apply the stock change (i.e. the stock changes on the right side of the time interval in which the losses apply)
+            next_stock = s * e + d
+        elif how == "linear":
+            # Assume the change happens at a constant rate, leading to a linear stock change, and exponential decay
+            next_stock = s * e + d * (e - 1) / math.log(e)
```

Comment:
The existing stock is already exponentially decaying, because with each consecutive (constant-resolution) time step, some constant percentage of the previous stock remains. For example:
```
100 -> 90 -> 81 -> 72.9 -> ...
```
For computing the decay of the deltas, we have to make an assumption on how a delta was distributed within a given time step. In case it happened at a constant rate, this leads to a linear stock change from one time step to the next.

And an $e$ is introduced when you apply exponential decay to that. To see that, imagine we cut one interval in $n$ pieces (each with a stock change $\frac{d}{n}$), apply the efficiency to each piece $k$ (for the corresponding fraction of the time step $k/n$), and then take the limit $s\textrightarrow\infty$:
$$\lim_{n\to\infty} \sum_{k=0}^{n} \frac{d}{n} \eta^{k/n}$$
[which is](https://www.wolframalpha.com/input?i=Limit%5BSum%5Bp%5E%28k%2Fn%29%2Fn%2C+%7Bk%2C+0%2C+n%7D%5D%2C+n+-%3E+Infinity%5D&assumption=%22LimitHead%22+-%3E+%7B%22Discrete%22%7D):
$$\frac{\eta - 1}{e^\eta}$$

Reply:
Okay, thanks. Maybe add "within the current interval" at the end of the comment, that could help another first-time reader to grasp this right away.